module #3. REST API Advanced/src/main/java/com/epam/esm/controller/constraint/FilterValidator.java

```diff
@@ -0,0 +1,56 @@
+package com.epam.esm.controller.constraint;
+
+import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;
+import org.springframework.util.MultiValueMap;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+public class FilterValidator implements ConstraintValidator<FilterConstraint, MultiValueMap<String, String>> {
+
+    private static final String INVALID_DATE_SORT = "40007";
+
+    private static final String INVALID_NAME_SORT = "40008";
+
+    private static final String ASCENDING = "asc";
+
+    private static final String DESCENDING = "desc";
+
+    @Override
+    public boolean isValid(MultiValueMap<String, String> stringStringMultiValueMap,
```

Comment:
This method is too complex, could you rewrite it?

Reply:
Is this ok?
```
    @Override
    public boolean isValid(MultiValueMap<String, String> filterParams,
                           ConstraintValidatorContext context) {

        return filterParams.toSingleValueMap()
                .entrySet()
                .stream()
                .filter(entry -> DATE_SORT_KEY.equals(entry.getKey()) || NAME_SORT_KEY.equals(entry.getKey()))
                .filter(entry -> !validateSortValue(context, entry.getValue()))
                .count() == 0; // can't use noneMatch as it leaves other values unchecked
    }

    boolean validateSortValue(ConstraintValidatorContext context, String value) {
        if (!(value.equals(ASCENDING) || value.equals(DESCENDING))) {
            context.disableDefaultConstraintViolation();
            context.buildConstraintViolationWithTemplate(SORT_ERROR_CODE)
                    .addPropertyNode(value)
                    .addConstraintViolation();
            return false;
        }
        return true;
    }
```